### PR TITLE
only allow adding unused columns

### PR DESCRIPTION
**Pull request by machisuji**
Work in Progress

WP [#19553](https://community.openproject.org/work_packages/19553)

This probably warrants a test. I will need at least another day for that, though.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://machisuji-fix-duplicate-available-columns.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-fix-duplicate-available-columns.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by machisuji** [comment]
This breaks rendering of added columns. I don't see the connection yet as I have no idea when it comes to anything in the frontend, specifically Angular. Also debugging angular is no fun.

screen shot 2015-04-13 at 17 29 49

The screenshot shows on the right the added 'Project' column whose heading and values cannot be rendered due to the errors seen on in the debugger.

**Comment by machisuji** [comment]
It does fix the column selection, though! :D

File: frontend/app/services/query-service.js [hunk 1]
```diff
@@ -164,10 +164,11 @@ module.exports = function(Query, Sortation, $http, PathHelper, $q, AVAILABLE_WOR
 
     loadAvailableUnusedColumns: function(projectIdentifier) {
       return QueryService.loadAvailableColumns(projectIdentifier)
-        .then(function(available_columns) {
-          availableUnusedColumns = WorkPackagesTableHelper.getColumnDifference(available_columns, QueryService.getSelectedColumns());
-          return availableUnusedColumns;
-        });
+        .then(QueryService.selectUnusedColumns);
```
Comment (floriank) [inline frontend/app/services/query-service.js]: if this is only used within the service, it should not be exposed in the service. could you just add the function to the outer scope, even outside the `QueryService` definition?

**Comment by NobodysNightmare** [comment]
@machisuji Are you still working on this? Are you road blocked here?

**Comment by machisuji** [comment]
Stepped away from the problem for a bit and coming back to it I see it clearly. I didn't see or consider the [side effects](https://github.com/opf/openproject/blob/dev/frontend/app/services/query-service.js#L168). Anyway here we go. This fixes it for real this time.

File: frontend/app/work_packages/controllers/dialogs/columns.js [hunk 2]
```diff
@@ -41,7 +41,7 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
   // Available selectable Columns
   vm.promise = QueryService.loadAvailableColumns()
     .then(function(availableColumns){
-      vm.availableColumns = availableColumns;
+      vm.availableColumns = QueryService.getAvailableUnusedColumns();
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/dialogs/columns.js]: Still have to ask:

The code reads: 
1. load available columns
2. then disregard those columns (given as parameter)
3. and take available unused columns instead

I guess the answer will be something like:
`loadAvailableColumns` does the http request, while `getAvailableUnusedColumns` does no networking, but returns the right thing...

Still it looks weird, to have this single most return value of `loadAvailableColumns` ignored...

Comment (machisuji) [inline frontend/app/work_packages/controllers/dialogs/columns.js]: It's not ignored. Read on and [you will see](https://github.com/machisuji/openproject/blob/fix/duplicate_available_columns/frontend/app/work_packages/controllers/dialogs/columns.js#L45) that it's actually still used.

**Comment by NobodysNightmare** [comment]
Arg...

Sorry @machisuji but this is not the fix.

Yes, I can't add e.g. the ID column if it is already added.
But after reomving a column from the selection, I still can't add it again!

### Reproduction
1. open the column selection
2. remove any column that is currently included
3. try to re-add that column

**Comment by machisuji** [comment]
@NobodysNightmare Fixed the issue you mentioned. The problem was that the columns available for selection were only updated when the modal was opened, not when the selection changes.

@myabc Unfortunately I implemented this with the refresh attribute of the select element. This is called with every key stroke the user makes when searching for a column. Ideally this would only happen when the selection changes (column removed, added), though. I couldn't get that to work, though, unfortunately (tried `ng-update` and `update` to no avail). Do you know how to do this?

All I need to happen is for the function `#updateUnusedColumns(selectedColumns)` to be called every time the selection (i.e. a column is added or removed from the selection) changes.

File: frontend/app/work_packages/controllers/dialogs/columns.js [hunk 3]
```diff
@@ -35,13 +35,16 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
   vm.selectedColumns = [];
   vm.oldSelectedColumns = [];
   vm.availableColumns = [];
+  vm.unusedColumns = [];
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/dialogs/columns.js:38]: What is the semantic difference between an available column and an unused column? Are both lists equal?

File: frontend/app/work_packages/controllers/dialogs/columns.js [hunk 4]
```diff
@@ -55,10 +58,12 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
     return _.difference(vm.selectedColumns, vm.oldSelectedColumns);
   }
 
+  function getColumnName(column) {
+    return column.name;
+  }
+
   function getColumnNames(arr) {
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/dialogs/columns.js:65]: Arrrrrr!

Arrrr!

File: frontend/app/work_packages/controllers/dialogs/columns.js (hunk 3 above)
Comment (machisuji) [inline frontend/app/work_packages/controllers/dialogs/columns.js:38]: Available columns are all columns you could ever select. Unused columns are those that are not currently selected.

So basically: `unusedColumns = availableColumns - selectedColumns`

**Comment by machisuji** [comment]
While the trick with the refresh is not the prettiest it is a valid solution and it works just fine. So unless I find out how to get notified of choice updates we can just use this IMO.

**Comment by NobodysNightmare** [comment]
Looks like a valid solution... If someone dislikes it, then he/she will have to make a new PR ;-P

File: frontend/app/work_packages/controllers/dialogs/columns.js [hunk 5]
```diff
@@ -55,10 +58,12 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
     return _.difference(vm.selectedColumns, vm.oldSelectedColumns);
   }
 
+  function getColumnName(column) {
```
Comment (floriank) [inline frontend/app/work_packages/controllers/dialogs/columns.js:61]: I do not get the Mehrwert of this method.

Comment (machisuji) [inline frontend/app/work_packages/controllers/dialogs/columns.js:61]: It saved me from writing the same inline function - which looks horrible in overly verbose Javascript - twice.
Probably saved me effectively a line while making the code more readable IMO.

**Pull request merged**